### PR TITLE
`Development`: Move assessment due date check into ExerciseDateService

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
Before the assessment due date, the student should under no circumstances see the manual tutor feedback. However, not every endpoint is currently checking and enforcing this rule.

### Description
This Refactoring is the first step to consistently check and remove scores and feedback on the server side if the student is not yet allowed to see it.

This PR adds the after assessment due date checks to the `ExerciseDateService` and uses these methods instead of always directly comparing the date with `now()`. This also makes the code more readable, since you now have a descriptive method name in these places and don't need to think yourself about the condition and what is meant by it.

### Steps for Testing
This PR requeires cafeull testsing and code review!

Please go through the code and check that the logic ist still correct (and was not unintentionally inverted).

1. Create exercises (every type) and participate in them as a student.
2. Grade the submissions as a tutor.
3. Check different places on Artemis (e.g. exercise detail, exercise overview, statistics, front page, etc.). Check that the student is never able to see the result generated by the tutor (before the assessment due date).
4. After the assessment due date, check the the result is now shown.

#### Exam Mode Testing
same as above, but with exam exercises. No results should be visible before the "publish exam results" date is reached.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test

### Test Coverage
added tests for ExerciseDateService (coverage unchanged since it's also covered by other integration tests)
Adjusted existing tests to check for filtering.

### CODE REVIEW

**Review by tobias-lippert — APPROVED** [review]
Carefully looked through the code and checked that the logic regarding due date checks wasn't accidentally inverted.
Didn't find any issues.

File: src/main/java/de/tum/in/www1/artemis/service/CourseScoreCalculationService.java [hunk 1]
```diff
@@ -255,8 +258,7 @@ private boolean includeIntoScoreCalculation(Exercise exercise) {
      * @param exercise the exercise whose assessment state should be determined
      */
     private boolean isAssessmentDone(Exercise exercise) {
-        boolean isNonAutomaticAssessmentDone = !isAssessedAutomatically(exercise)
-                && (exercise.getAssessmentDueDate() == null || exercise.getAssessmentDueDate().isBefore(ZonedDateTime.now()));
+        boolean isNonAutomaticAssessmentDone = !isAssessedAutomatically(exercise) && (exerciseDateService.isAfterAssessmentDueDate(exercise));
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/CourseScoreCalculationService.java:261]:
```suggestion
        boolean isNonAutomaticAssessmentDone = !isAssessedAutomatically(exercise) && exerciseDateService.isAfterAssessmentDueDate(exercise);
```

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 2]
```diff
@@ -214,6 +214,7 @@ public Result addFeedbackToResult(@NotNull Result result, List<Feedback> feedbac
      * @param result the result for which the feedback elements should be returned
      * @return the list of filtered feedbacks
      */
+    // TODO improve naming (more like prepareFeedbackForClient)
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/ResultService.java:217]: Why don't you resolve this todo?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by b-fein — DISMISSED** [review]
Code changes look good.

**Review by stefanwaldhauser — DISMISSED** [review]
Sensible refactoring

**Review by JohannesStoehr — DISMISSED** [review]
Thanks for addressing my review. Code looks good to me

**Review by TimOrtel — DISMISSED** [review]
The code changes look fine. I think its a very nice refactoring to eliminate duplicated code.

**Review by b-fein — DISMISSED** [review]
re-:+1: code

**Review by JohannesStoehr — COMMENTED** [review]
Benedict made some methods static to reduce the dependencies between the services in #5842. Maybe we can do the same with the new methods here?

**Review by JohannesStoehr — DISMISSED** [review]
Comments were addressed. Code looks good to me

**Review by b-fein — DISMISSED** [review]
re-:+1: code

**Pull request closed without merging**